### PR TITLE
chore(cheatcodes): ArbitraryStorage as option

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
re https://github.com/foundry-rs/foundry/pull/8807#discussion_r1755012005

### CODE REVIEW

**Comment by grandizzy** [comment]
@DaniPopes would making `ArbitraryStorage` as an Option and check if some in step_end be better than checking the op::SLOAD? pls check, thank you

**Comment by DaniPopes** [comment]
Yeah that makes sense

File: crates/cheatcodes/src/inspector.rs [hunk 1]
```diff
@@ -1080,6 +1070,28 @@ impl Cheatcodes {
             None => StdRng::from_entropy(),
         })
     }
+
+    /// Returns existing or set a default `ArbitraryStorage` option.
+    /// Used by `setArbitraryStorage` cheatcode to track addresses with arbitrary storage.
+    pub fn ensure_arbitrary_storage(&mut self) -> &mut ArbitraryStorage {
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:1076]: can just be `fn arbitrary_storage`

File: crates/cheatcodes/src/inspector.rs [hunk 2]
```diff
@@ -1080,6 +1070,28 @@ impl Cheatcodes {
...
+        self.arbitrary_storage.get_or_insert(ArbitraryStorage::default())
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:1077]:
```suggestion
        self.arbitrary_storage.get_or_insert_with(ArbitraryStorage::default)
```

File: crates/cheatcodes/src/inspector.rs [hunk 3]
```diff
@@ -1577,25 +1586,28 @@ impl Cheatcodes {
...
-                        arbitrary_value,
-                    );
-                } else {
-                    self.arbitrary_storage.save(
-                        &mut ecx.inner,
-                        target_address,
-                        key,
-                        arbitrary_value,
-                    );
+        if interpreter.current_opcode() == op::SLOAD {
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:1589]: can early return on this and sload below with `let ... else`

File: crates/cheatcodes/src/utils.rs [hunk 4]
```diff
@@ -162,15 +162,19 @@ impl Cheatcode for setArbitraryStorageCall {
 impl Cheatcode for copyStorageCall {
     fn apply_stateful<DB: DatabaseExt>(&self, ccx: &mut CheatsCtxt<DB>) -> Result {
         let Self { from, to } = self;
+
         ensure!(
-            !ccx.state.arbitrary_storage.is_arbitrary(to),
+            !&ccx.state.has_arbitrary_storage(to),
```
Comment (DaniPopes) [inline crates/cheatcodes/src/utils.rs:167]:
```suggestion
            !ccx.state.has_arbitrary_storage(to),
```

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/cheatcodes/src/inspector.rs (hunk 3 above)
Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs]: added in https://github.com/foundry-rs/foundry/pull/8848/commits/aec556840a846074521b048271983e44906ef7a8

File: crates/cheatcodes/src/inspector.rs (hunk 1 above)
Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:1076]: this one needs to be public as it it called when `setArbitraryStorage` to set option to some, can rename to `arbitrary_storage()`

https://github.com/foundry-rs/foundry/blob/aec556840a846074521b048271983e44906ef7a8/crates/cheatcodes/src/utils.rs#L153-L160

Comment (DaniPopes) [reply crates/cheatcodes/src/inspector.rs:1076]: yes i mean rename

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs]: ah, kk, changed in https://github.com/foundry-rs/foundry/pull/8848/commits/0490298d37f4d9f040e05890a810f29e51341363

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**